### PR TITLE
[SYCLomatic] Fix some bugs for migrating APPs using Pytorch as a library

### PR DESCRIPTION
**Pull request by zhiweij1**
1. Fix dpct crash when assertion is enabled.
2. Migrate header in analysis scope but not in the in-root
3. Fix inserting headers of spBLAS
4. Insert BLAS header for cuBLAS types
5. Fix some Pytorch framework API migration rules

Signed-off-by: Jiang, Zhiwei <zhiwei.jiang@intel.com>

### CODE REVIEW

File: clang/lib/DPCT/RulesLang/RulesLang.cpp [hunk 1]
```diff
@@ -34,6 +34,7 @@
 #include "clang/Lex/MacroArgs.h"
 #include "llvm/ADT/SCCIterator.h"
 #include "llvm/ADT/SmallString.h"
+#include "llvm/ADT/StringRef.h"
```
Comment (zhiweij1) [inline clang/lib/DPCT/RulesLang/RulesLang.cpp]: remove

File: clang/lib/DPCT/RulesLang/RulesLang.cpp [hunk 2]
```diff
@@ -857,6 +858,7 @@ void TypeInDeclRule::runRule(const MatchFinder::MatchResult &Result) {
       }
     }
 
+
```
Comment (zhiweij1) [inline clang/lib/DPCT/RulesLang/RulesLang.cpp]: remove

File: clang/lib/DPCT/RulesInclude/InclusionHeaders.cpp [hunk 3]
```diff
@@ -200,7 +200,8 @@ void IncludesCallbacks::InclusionDirective(
         DpctGlobalInfo::getIncludeMapSet().push_back({IncludedFile, Repl});
       }
     }
-    return;
+    if (Global.isInRoot(IncludedFile))
```
Comment (zhiweij1) [inline clang/lib/DPCT/RulesInclude/InclusionHeaders.cpp:203]: add test

**Review by tomflinda — APPROVED** [review]
LGTM

**Review by ziranzha — APPROVED** [review]

**Pull request merged**